### PR TITLE
Feature tip statistics

### PR DESCRIPTION
**Pull request by DarthRevan77797**

### CODE REVIEW

File: src/main/java/umk/mat/pajda/ProjektZespolowy/repository/TipRepository.java [hunk 1]
```diff
@@ -1,8 +1,31 @@
...
+
+  @Query("select count(*) from Tip t where t.user.id = :userid")
+  Integer getNumberOfTips(@Param("userid") Integer userid);
+
+  @Query(
+      "select new umk.mat.pajda.ProjektZespolowy.DTO.TipMonthDTO(SUM(t.amount),"
+          + "EXTRACT(month from t.createdAt),EXTRACT(year from t.createdAt),t.currency) "
+          + "from Tip t where t.user.id = :userid "
+          + "group by EXTRACT(month from t.createdAt), EXTRACT(year from t.createdAt) "
+          + "order by EXTRACT(year from t.createdAt) desc,EXTRACT(month from t.createdAt) desc")
```
Comment (ThePapiok) [inline src/main/java/umk/mat/pajda/ProjektZespolowy/repository/TipRepository.java:29]: I think it need tests you can also add tests to other  custom queries

Comment (DarthRevan77797) [reply src/main/java/umk/mat/pajda/ProjektZespolowy/repository/TipRepository.java:29]: done

File: src/main/java/umk/mat/pajda/ProjektZespolowy/services/TipService.java [hunk 2]
```diff
@@ -95,4 +101,26 @@ public ResponseEntity<String> createPayment(OpinionPostDTO opinionPostDTO, Strin
...
+      TipStatisticsGetDTO tipStatisticsGetDTO = new TipStatisticsGetDTO();
+      Tip tipMin = tipRepository.findFirstByUserOrderByAmountDesc(user);
+      tipStatisticsGetDTO.setMaxTipAmount(tipMin.getAmount());
+      tipStatisticsGetDTO.setCurrency(tipMin.getCurrency());
+
+      tipStatisticsGetDTO.setMinTipAmount(
+          tipRepository.findFirstByUserOrderByAmountAsc(user).getAmount());
+      tipStatisticsGetDTO.setAvgTipAmount(tipRepository.getAvgAmountForAllTips(user.getId()));
+      tipStatisticsGetDTO.setSumTipValueForEveryMonth(
+          tipRepository.getSumAmountForEachMonth(user.getId()));
```
Comment (ThePapiok) [inline src/main/java/umk/mat/pajda/ProjektZespolowy/services/TipService.java:117]: i get error for user with id 2

File: src/main/java/umk/mat/pajda/ProjektZespolowy/services/TipService.java [hunk 3]
```diff
@@ -95,4 +101,26 @@ public ResponseEntity<String> createPayment(OpinionPostDTO opinionPostDTO, Strin
...
     this.restTemplate = restTemplate;
   }
+
+  public TipStatisticsGetDTO getStatistics(String userName) {
+    try {
+      User user = userRepository.findByMail(userName).get();
+      TipStatisticsGetDTO tipStatisticsGetDTO = new TipStatisticsGetDTO();
+      Tip tipMin = tipRepository.findFirstByUserOrderByAmountDesc(user);
+      tipStatisticsGetDTO.setMaxTipAmount(tipMin.getAmount());
+      tipStatisticsGetDTO.setCurrency(tipMin.getCurrency());
```
Comment (ThePapiok) [inline src/main/java/umk/mat/pajda/ProjektZespolowy/services/TipService.java:111]: What if client will pay onece in PLN and then in USD

File: src/main/java/umk/mat/pajda/ProjektZespolowy/services/TipService.java (hunk 2 above)
Comment (DarthRevan77797) [reply src/main/java/umk/mat/pajda/ProjektZespolowy/services/TipService.java:117]: fixed

Comment (ThePapiok) [reply src/main/java/umk/mat/pajda/ProjektZespolowy/services/TipService.java:117]: image
In a given month, should the sum together regardless of currency?

Comment (DarthRevan77797) [reply src/main/java/umk/mat/pajda/ProjektZespolowy/services/TipService.java]: resolved , implementation change

File: src/main/java/umk/mat/pajda/ProjektZespolowy/services/TipService.java (hunk 3 above)
Comment (DarthRevan77797) [reply src/main/java/umk/mat/pajda/ProjektZespolowy/services/TipService.java]: resolved , implementation change

**Review by ThePapiok — APPROVED** [review]
nice

**Pull request merged**